test/langtools/tools/javac/nullability/NullabilitySignatureAttrTests.java

```diff
@@ -0,0 +1,345 @@
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8338910
+ * @summary [lw5] enhance the Signature attribute to represent nullability
+ * @library /lib/combo /tools/lib
+ * @modules
+ *     jdk.compiler/com.sun.tools.javac.util
+ *     jdk.compiler/com.sun.tools.javac.api
+ *     jdk.compiler/com.sun.tools.javac.main
+ *     jdk.compiler/com.sun.tools.javac.code
+ *     jdk.jdeps/com.sun.tools.classfile
+ * @build toolbox.ToolBox toolbox.JavacTask
+ * @run junit NullabilitySignatureAttrTests
+ */
+
+import java.io.File;
+
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+import java.util.List;
+import java.util.Set;
+
+import com.sun.tools.javac.util.Assert;
+
+import com.sun.tools.classfile.Attribute;
+import com.sun.tools.classfile.Attributes;
+import com.sun.tools.classfile.ClassFile;
+import com.sun.tools.classfile.Code_attribute;
+import com.sun.tools.classfile.ConstantPool;
+import com.sun.tools.classfile.ConstantPool.CONSTANT_Class_info;
+import com.sun.tools.classfile.ConstantPool.CONSTANT_Fieldref_info;
+import com.sun.tools.classfile.ConstantPool.CONSTANT_Methodref_info;
+import com.sun.tools.classfile.ImplicitCreation_attribute;
+import com.sun.tools.classfile.NullRestricted_attribute;
+import com.sun.tools.classfile.Field;
+import com.sun.tools.classfile.Instruction;
+import com.sun.tools.classfile.Method;
+import com.sun.tools.classfile.Signature_attribute;
+
+import org.junit.jupiter.api.Test;
+import tools.javac.combo.CompilationTestCase;
+
+import toolbox.*;
+import toolbox.Task.*;
+
+public class NullabilitySignatureAttrTests extends CompilationTestCase {
+    ToolBox tb = new ToolBox();
+
+    private static String[] PREVIEW_OPTIONS = {"--enable-preview", "-source", Integer.toString(Runtime.version().feature())};
+
+    public NullabilitySignatureAttrTests() {
+        setDefaultFilename("Nullable.java");
+        setCompileOptions(PREVIEW_OPTIONS);
+    }
+
+    record SignatureData(String source, String expectedSignature) {}
+
+    final List<SignatureData> signatureDataList = List.of(
+            new SignatureData(
+                    """
+                    class Test {
+                        Test! t;
+                    }
+                    """,
+                    "LTest!;"
```

Comment:
My feeling is that the `!`/`?` should go _before_ the type in the sig attribute, similar to what we do for arrays and wildcards. Otherwise parsing becomes more complex.